ADR/ADR-003-PyBind11-For-CPP-Bindings.md

```diff
@@ -0,0 +1,295 @@
+  increases the cognitive load and makes onboarding more difficult for new
+  developers.
+
+
+### Business Challenges
+
+- Slower Time-to-Prodcution: The complexity of maintaining CFFI-based
+  bindings delays the delivery of new features and tools to users. 
+- Inconsistent User Experience: Manually wrapped APIs often lack the polish
+  and consistency expected by Python users, affecting usability and
+  adoption.
+- Scalability Issues: As ECMWF’s software ecosystem grows, the current
+  approach does not scale well in terms of maintainability, performance,
+  and developer productivity.
+
+
+### Goal
+
+To adopt a modern, maintainable, and performant solution for exposing C++ code to Python that:
+
+- Supports direct binding of C++ constructs without intermediate C layers
+- Reduces boilerplate and maintenance effort
+- Integrates smoothly with ECMWF’s build and packaging systems
+- Enables faster development cycles and better user experience
+
+
+### Options Considered
+
+#### Manual Binding with Python C API (CPython)
+
+[source](https://github.com/python/cpython) | [documentation](https://docs.python.org/3/c-api/index.html)
+
+Licence: Python Software Foundation License (PSFL), BSD-style, GPL-compatible
+Maturity: Start in 1991 with python
+
+This approach uses the official Python C API (<Python.h>) to directly expose C++ functions or classes to Python. 
+
+The workflow is:
+  1. Define C++ Code:
+     Write normal C++ headers and implementation files for the logic you want to expose (e.g., mylib.h, mylib.cpp).
+
+  2. Create Binding Layer:
+     Implement a small C++ file that:
+
+       * Includes <Python.h> and your C++ headers
+       * Wraps C++ functions using `PyArg_ParseTuple` (to read Python arguments) 
+         and `Py_BuildValue/PyLong_FromLong` (to return values)
+
+       * Registers methods in a `PyMethodDef` table
+         Provides a `PyInit_<module>` function to initialize the Python module
+
+  3. Compile to Shared Library:
+     Build the code into a shared object (.so/.pyd) using the Python development headers and link flags. 
+     Python can then import this module directly.
+
+  4. Usage in Python:
+     Once built, the module can be imported like any native Python package (import mymodule) 
+     and the wrapped C++ functions can be called as normal Python functions.
+
+
+#### CFFI
+
+[source](https://github.com/python-cffi/cffi) | [documentation](https://cffi.readthedocs.io/en/stable/)
+
+Licence: MIT
+Maturity: Start in 2012, many releases (33 tags on github), good documentation 
+Longevity: 8 maintainers (pypy), 77 contributors (github)
+Activity: Multiple releases in the last years, community interaction
+Visibility: 204 stars on Github 
+
+CFFI (C Foreign Function Interface) is a Python library that allows Python code
+to call C functions and use C data types. With CFFI it is possible to
+ 
+ * declare C interfaces: write C declarations (function prototypes, structs,
```

Comment:
Note that this typically will just be a (post-processed) header files for the primary API - we don't need to write declarations just for this.